Add cloud provider guides (AWS, Azure, GCP) (#829)
memgraph/documentation: pages/deployment/aws.mdx

---
title: Deploy Memgraph on AWS EC2 instance 
description: Learn how to deploy your Memgraph instance on the AWS EC2 instance.
---

import { Card, Cards, Tabs, Tab } from 'nextra/components'

> # Deploy Memgraph in AWS EC2 instance

Reviewer: in AWS instance or on AWS instance (I saw you used 'on' in other places)

Author: ```suggestion
# Deploy Memgraph on AWS EC2 instance
```

This guide will show you how to deploy Memgraph on an AWS EC2 instance. The
guide will cover only the specific bits that are different from the general

> deployment guide you can find for native [linux](./linux.mdx) or
> [docker](./docker.mdx) deployments.

Reviewer: ```suggestion
deployment guide you can find for native [Linux](./linux.mdx) or
[Docker](./docker.mdx) deployments.
```

This guide assumes you have an AWS account and are familiar with the AWS
console. If you are not, you can follow the [AWS EC2
documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EC2_GetStarted.html).

> ## Crating a new EC2 instance 

Reviewer: ```suggestion
## Creating a new EC2 instance 
```

The first step is to have an EC2 instance running on AWS, where you will deploy

> Memgraph. Here are some guidelines for opening a new EC2 instance for memgraph. 

Reviewer: ```suggestion
Memgraph. Here are some guidelines for opening a new EC2 instance for Memgraph. 
```

### Picking the OS 

When creating the EC2 instance, you need to pick the OS you wish to have on the
instance. If you are going to use the Memgraph native installation, you need to
pick a Linux distribution that Memgraph supports.

Memgraph supports the following Linux distributions:

- Ubuntu (20.04, 22.04)
- Fedora (38, 39)
- Debian (11, 12) 
- Centos (7, 9) 
- Amazon Linux (2)
- RedHat (7, 9)

> - Rocky (9.3)

Reviewer: I would be nice to have this listed somewhere else (e.g. Installation), so it can be updated only on that page in the future (and not on every one where it's mentioned). @kgolubic Do we have such page or section where all supported Linux distributions are listed?

Reviewer 2: @katarinasupe we have a page with direct download links -> https://memgraph.com/docs/getting-started/install-memgraph/direct-download-links. If needed I can create a such page, or simlply add a paragraph to the top of direct download page that these are the supported Linux distributions.

Author: I have updated this

This way, you will manage the native installation of Memgraph, which can bring
some of the speed benefits. 

If you are going to use the Memgraph Docker image, you can pick any Linux
distribution you prefer, as long as you can run Docker on it.

### Picking the instance type

When creating the EC2 instance, you need to pick the instance type. The process
highly depends on your actuall workload. For heavy graph processing, consider
using instances with more CPU and memory. 

> Memgraph, by default, stores all data in the RAM, if you are going to use the
> default storage mode (`IN_MEMORY_TRANSACTIONAL` or `IN_MEMORY_ANALYTICAL`), make
> a calculation [how much data you will
> need]((/fundamentals/storage-memory-usage#calculate-storage-memory-usage)).

Reviewer: ```suggestion
If you are running Memgraph in `IN_MEMORY_TRANSACTIONAL` (default mode) or in `IN_MEMORY_ANALYTICAL` storage mode, all data is stored in RAM. That means it would be good to first calculate [how much memory you will
need](/fundamentals/storage-memory-usage#calculate-storage-memory-usage).
```

There is also an [easy-to-use
calulator](https://memgraph.com/enterprise#ram-calculator) for approximating

> memory usage. The rule of thumb is to have double the memory of what your

Reviewer: ```suggestion
memory usage. The good rule of thumb is to have double the memory of what your
```

storage would take. You can go with less if you do not have a demanding
workload.

> AWS has specialized instances for [memory-optimized
> workloads](https://docs.aws.amazon.com/ec2/latest/instancetypes/mo.html) wich

Reviewer: ```suggestion
AWS has specialized instances for [memory-optimized
workloads](https://docs.aws.amazon.com/ec2/latest/instancetypes/mo.html) which
```

are designed for in-memory databases, data analytics, and other memory-intensive

> applications. They also provide DDR5 memory, so you may consider using them to
> optimize for better performance. 

Reviewer: ```suggestion
applications. They also provide DDR5 memory, so you may consider using such instances to
optimize for better performance. 
```

### Setup the network

When creating the EC2 instance, you need to set up the network access and

> security group. By default, the Memgraph uses port 7687 for the Bolt protocol.

Reviewer: ```suggestion
security group. By default, Memgraph uses port 7687 for the Bolt protocol.
```

You need to open this port for TCP traffic and to allow connections to Memgraph. 

If you change the default bolt port, make sure to update the security group
accordingly.

> Also, if deploying the Memgraph for replication or clustering, the ports for the
> replication and clustering should be open as well.

Reviewer: ```suggestion
Also, if deploying Memgraph for replication or clustering, the ports for the
replication and clustering should be open as well.
```

I guess replication here is community, clustering is referring to HA?

### Setup the storage 

When creating the EC2 instance, you need to set up the storage. By default
Memgraph stores all data to working RAM, but for the [persistency between
restarts](../fundamentals/data-durability) Memgraph uses the disk storage to
store snapshots, configurations, etc.

The sizing of the storage depends on the amount of data you are going to store
in your Memgraph instance, and the quantity of the snapshots you want to keep
alive. 

The general rule of thumb is to double the storage and then the amount of data
you will store. If you are going to store 100GB of data, you should have at

> least 200GB of disk storage. 

Author: I have rewritten this.

## Installing Memgraph 

Depending on the way you want to deploy Memgraph, native or via Docker, you need
to follow the steps below: 

<Tabs items={["Docker" , "Linux"]}>

 <Tab> 
    For example, if you are using Amazon Linux 2, you can follow the steps
    below to install Docker:

        ```bash
            sudo yum update -y
            sudo yum install docker -y
            sudo systemctl start docker
            # Add the user to the docker group
            sudo usermod -a -G docker $(whoami)
            newgrp docker
        ```
    Note that the steps above may vary depending on your Linux distribution.

>     After the docker is installed, you can pull the Memgraph image and run it:

Reviewer: ```suggestion
    After Docker is installed, you can pull the Memgraph image and run it:
```

        ```bash
            docker run -p 7687:7687 memgraph/memgraph:latest

>         ```

Reviewer: This is unnecessarily indented. (check deployment)

Author: Done.

    This will run Memgraph on the default port 7687. You should be able to connect
        to it via [Memgraph Lab](../data-visualization) or via [client

>     libraries](../client-libraries) using the bolt protocol. If you have an issue

Reviewer: ```suggestion
    libraries](../client-libraries) using the Bolt protocol. If you have an issue
```

    connecting to Memgraph remotely, make sure that the port 7687 is open in [the

>     security group of your EC2 instance](#Setup-the-network).

Reviewer: ```suggestion
    security group of your EC2 instance](#setup-the-network).
```

  </Tab>

    <Tab>

    Depending on the Linux distribution you are using, you can follow the guide for
    installing Memgraph on each distribution: 

        - [Ubuntu](../getting-started/install-memgraph/ubuntu.mdx)
        - [Fedora](../getting-started/install-memgraph/fedora.mdx)
        - [Debian](../getting-started/install-memgraph/debian.mdx)
        - [Centos](../getting-started/install-memgraph/centos.mdx)
        - [Amazon Linux](../getting-started/install-memgraph/amazon-linux.mdx)
        - [RedHat](../getting-started/install-memgraph/redhat.mdx)
        - [Rocky](../getting-started/install-memgraph/rocky.mdx)

    </Tab> 
 </Tabs>

## Manage Memgraph deployment

After Memgraph is installed and running on your EC2 instance, Memgraph
management in AWS is identical to the general guidelines that are described in
the form of a native Linux Memgraph or Docker container Memgraph. 

Depending on what you are using, you can follow the [Linux](./linux.mdx) or
[Docker](./docker.mdx) deployment guide for more information on how to manage
the Memgraph deployment.

> # Next steps 

Reviewer: ```suggestion
## Next steps 
```

---
title: Deploy Memgraph on AWS EC2 instance 

> description: Learn how to deploy your Memgraph instance on the AWS EC2 instance.

Reviewer 3: Avoid defining a term by repeating the term itself. 

In this case, it is better to give a thumbnail sketch of the content, such as choosing instance type as memory optimized, selecting the OS, network considerations, etc.

Author: Good point.

guide will cover only the specific bits that are different from the general
deployment guide you can find for native [Linux](./linux.mdx) or
[Docker](./docker.mdx) deployments.

This guide assumes you have an AWS account and are familiar with the AWS
console. If you are not, you can follow the [AWS EC2
documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EC2_GetStarted.html).

## Creating a new EC2 instance 

The first step is to have an EC2 instance running on AWS, where you will deploy
Memgraph. Here are some guidelines for opening a new EC2 instance for Memgraph. 

### Picking the OS 

When creating the EC2 instance, you need to pick the OS you wish to have on the
instance. If you are going to use the Memgraph native installation, you need to
pick a Linux distribution that Memgraph supports.

Memgraph supports multiple Linux distributions. Memgraph packages for supported distributions and versions can be downloaded 
from the [following page](/getting-started/install-memgraph/direct-download-links#linux)

This way, you will manage the native installation of Memgraph, which can bring
some of the speed benefits. 

> If you are going to use the Memgraph Docker image, you can pick any Linux

Reviewer 3: s/pick any Linux/select whichever Linux/;

distribution you prefer, as long as you can run Docker on it.

> 

Reviewer 3: We don't mention anything about choosing an architecture (x86 or arm). Please add that.

### Picking the instance type

When creating the EC2 instance, you need to pick the instance type. The process
highly depends on your actuall workload. For heavy graph processing, consider

> using instances with more CPU and memory. 

Reviewer 3: Please provide the user with better guidance in how to make this choice. Specifically, we should steer them towards memory optimized such as the R5 family or X1 family.

Reviewer 3: Combine this paragraph with the one below (`AWS has specialized...`).

If you are running Memgraph in `IN_MEMORY_TRANSACTIONAL` (default mode) or in `IN_MEMORY_ANALYTICAL` storage mode, all data is stored in RAM. That means it would be good to first calculate [how much memory you will
need](/fundamentals/storage-memory-usage#calculate-storage-memory-usage).
There is also an [easy-to-use
calulator](https://memgraph.com/enterprise#ram-calculator) for approximating
memory usage. The good rule of thumb is to have double the memory of what your
storage would take. You can go with less if you do not have a demanding
workload.

AWS has specialized instances for [memory-optimized
workloads](https://docs.aws.amazon.com/ec2/latest/instancetypes/mo.html) which
are designed for in-memory databases, data analytics, and other memory-intensive
applications. They also provide DDR5 memory, so you may consider using such instances to

> optimize for better performance. 

Reviewer 3: Actually we aren't optimizing for better performance. Rather we are optimizing for cost per GB of RAM.

Author: Yep, that is a great point. AWS always mentions performance as a leading metric, but the storage cost is lower, so it is a win-win.

> ### Setup the network

Reviewer 3: Do we need to talk about VPN access? Or are we assuming that all access will be purely within the cloud?

When creating the EC2 instance, you need to set up the network access and
security group. By default, Memgraph uses port 7687 for the Bolt protocol.

> You need to open this port for TCP traffic and to allow connections to Memgraph. 

Reviewer 3: Link to the AWS docs about VPC security groups.

Author: Done

If you change the default Bolt port, make sure to update the security group
accordingly.

Also, if deploying Memgraph for [Replication](/clustering/replication#replication) or [High
Availability](/clustering/high-availability#high-availability-enterprise), the ports for the

> replication and clustering should be open as well.

Reviewer 3: List the default ports for replication and clustering

### Setup the storage 

When creating the EC2 instance, you need to set up the storage. By default
Memgraph stores all data to working RAM, but for the [persistency between
restarts](../fundamentals/data-durability) Memgraph uses the disk storage to
store snapshots, configurations, etc.

The sizing of the storage depends on the amount of data you are going to store
in your Memgraph instance, and the quantity of the snapshots you want to keep
alive. 

> 

Reviewer 3: Do we recommend EBS or local storage? Why? Does it matter if the storage is magnetic or SSD?

Author: Added a bit of color here.

The general rule of thumb is to double the storage and then the amount of data
you will store. If you are going to store 100GB of data, you should have at
least 200GB of disk storage. 

> ## Installing Memgraph 

Reviewer 3: Is this a duplicate of doc elsewhere on our site? If so, please point there instead.

Reviewer 3: I would drop them, since we aren't Docker -> we don't need to explain their product.

But! We must always be explicit in our assumptions about the user's environment.

Author: Deleted the docker part.

    security group of your EC2 instance](#setup-the-network).
  </Tab>

    <Tab>

    Depending on the Linux distribution you are using, you can follow the guide for
    installing Memgraph on each distribution: 

        - [Ubuntu](../getting-started/install-memgraph/ubuntu.mdx)
        - [Fedora](../getting-started/install-memgraph/fedora.mdx)
        - [Debian](../getting-started/install-memgraph/debian.mdx)
        - [Centos](../getting-started/install-memgraph/centos.mdx)
        - [Amazon Linux](../getting-started/install-memgraph/amazon-linux.mdx)
        - [RedHat](../getting-started/install-memgraph/redhat.mdx)
        - [Rocky](../getting-started/install-memgraph/rocky.mdx)

    </Tab> 
 </Tabs>

## Manage Memgraph deployment

After Memgraph is installed and running on your EC2 instance, Memgraph
management in AWS is identical to the general guidelines that are described in
the form of a native Linux Memgraph or Docker container Memgraph. 

Depending on what you are using, you can follow the [Linux](./linux.mdx) or
[Docker](./docker.mdx) deployment guide for more information on how to manage
the Memgraph deployment.

> ## Where to next? 

Reviewer 3: This should be more formal

Author: Can you provide a bit of color on this? Except for the Kubernetes paragraph at the top, it is mostly copy-passed from other places on docs, so maybe we should fix it in other places?

Reviewer 3: Oh. 

Yes, we should fix it in other places too, but that's a separate PR. Let's keep it consistent. I withdraw this comment at this time.
